Autopilot/Src/PID.cpp

```diff
@@ -1,78 +1,71 @@
 /**
  * @file PID.cpp
- * @author Ian Frosst
- * @date  March 26, 2017
- * @copyright Waterloo Aerial Robotics Group 2017 \n
- *   https://raw.githubusercontent.com/UWARG/ZeroPilot-SW/devel/LICENSE.md
+ * @authors Ian Frosst, Anthony Berbari
  */
 
-#include "PID.h"
+#include "PID.hpp"
 
-/**
- * Filtering constant for derivative. Between 0 and 1.
- * The larger this is, the more twitchy D control is.
- */
-#define FILTER (0.4f)
+/***********************************************************************************************************************
+ * Code
+ **********************************************************************************************************************/
 
-/* Generic PID functions. Can be used to PID other things (flaps, etc) */
+PIDController::PIDController(float _kp, float _ki, float _kd, float _i_max, float _min_output, float _max_output){
+	kp = _kp;
+	kd = _kd;
+	ki = _ki;
 
+	i_max = _i_max;
+	min_output = _min_output;
+	max_output = _max_output;
 
-PIDController::PIDController(float _kp, float _ki, float _kd, float _scale, int16_t _i_max){
-  kp = _kp;
-  kd = _kd;
-  ki = _ki;
-  i_max = _i_max;
-  scale = _scale;
-
-  integral = 0;
-  last_time = 0;
-  last_err = 0;
-  last_der = 0;
+	integral = 0.0f;
+	historicalValue[0] = 0.0f;
+	historicalValue[1] = 0.0f;
+	historicalValue[2] = 0.0f;
 }
 
-// PID loop function. error is (setpointValue - currentValue)
-float PIDController::PIDControl(float error) {
-  float output = 0;
+float PIDController::execute(float desired, float actual, float actualRate) {
 
-  uint64_t now = 0;  // getTimeUs(); TODO
-  uint32_t delta_usec = (now - last_time);
+	float error = desired - actual;
+	float derivative;
 
-  // check if we've gone too long without updating (keeps the I and D from
-  // freaking out)
-  if (delta_usec > PID_RESET_TIME || last_time == 0) {
-    delta_usec = 0;
-    integral = 0;
-    last_err = error;
-  }
+	integral +=  error;
```

Comment:
Would it make sense to add a flag to let whatever is calling the PID module to let it know that the system hasn't been updated for a while if this happens?

Reply:
I don't think so. This method only works properly if it's called at a constant interval (and the caller is warned about that in the .hpp file). If that doesn't happen, that's the caller's problem.